src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1429,6 +1442,51 @@ protected MembersAndInitializers GetMembersAndInitializers()
             return membersAndInitializers!;
         }
 
+        /// <summary>
+        /// The purpose of this function is to assert that the <paramref name="member"/> symbol
+        /// is actually among the symbols cached by this type symbol in a way that ensures
+        /// that any consumer of standard APIs to get to type's members is going to get the same 
+        /// symbol (same instance) for the member rather than an equvalent, but different instance.
+        /// </summary>
+        [Conditional("DEBUG")]
+        internal void AssertMemberExposure(Symbol member, bool forDiagnostics = false)
+        {
+            if (member is NamedTypeSymbol type)
+            {
+                Debug.Assert(forDiagnostics);
+                Debug.Assert(Volatile.Read(ref _lazyTypeMembers)?.Values.Any(types => types.Contains(t => t == (object)type)) == true);
+                return;
+            }
+            else if (member is TypeParameterSymbol || member is SynthesizedMethodBaseSymbol)
+            {
+                Debug.Assert(forDiagnostics);
+                return;
+            }
+            else if (member is FieldSymbol field && field.AssociatedSymbol is EventSymbol e)
+            {
+                Debug.Assert(forDiagnostics);
+                member = e;
```

Comment:
Not sure I understand this one. Can you add a comment to explain why we need to detour to the event?

Reply:
> Not sure I understand this one. Can you add a comment to explain why we need to detour to the event?

Fields like that are not added to members list. I can add a comment, but prefer doing that in a separate PR, unless there are other reasons to make changes in this one. 

---
In reply to: [571285867](https://github.com/dotnet/roslyn/pull/50961#discussion_r571285867) [](ancestors = 571285867)

Reply:
Perfectly fine as a followup.

---
In reply to: [571299314](https://github.com/dotnet/roslyn/pull/50961#discussion_r571299314) [](ancestors = 571299314,571285867)